File: rails/test/middleware_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require "json"

require "x402/middleware"
require "x402/platform_client"

class MiddlewareTest < Minitest::Test
  CFG = X402::Config.new(
    key_id: "key", secret: "sk", base_url: "https://platform.test", env: "production"
  )

  DOWNSTREAM = ->(_env) { [200, { "Content-Type" => "text/plain" }, ["paid content"]] }

  # PlatformClient wired to a canned (status, body) transport.
  def client(status, body)
    X402::PlatformClient.new(CFG, transport: ->(_u, _b, _h) { [status, body] })
  end

  # PlatformClient whose transport raises — simulates a network exception.
  def throwing_client
    X402::PlatformClient.new(CFG, transport: ->(_u, _b, _h) { raise "connection refused" })
  end

  def call(mw, env_overrides = {})
    env = { "PATH_INFO" => "/premium", "REQUEST_METHOD" => "GET" }.merge(env_overrides)
    mw.call(env)
  end

  def downstream_ran?(status, body)
    status == 200 && body == ["paid content"]
  end

  # --- no proof → /challenge ---

  def test_no_proof_challenge_200_relayed_as_402
    mw = X402::Middleware.new(DOWNSTREAM, route: "/premium",
                              client: client(200, '{"challenge":"pay-me"}'))
    status, _h, body = call(mw)
    assert_equal 402, status
    assert_includes body.join, "pay-me"
  end

  def test_no_proof_challenge_404_relayed_as_404_hardcoded
    mw = X402::Middleware.new(DOWNSTREAM, route: "/premium", client: client(404, "{}"))
    status, _h, body = call(mw)
    assert_equal 404, status
    assert_equal({ "error" => "no_such_route" }, JSON.parse(body.join))
  end

  def test_no_proof_challenge_500_fails_closed_502
    mw = X402::Middleware.new(DOWNSTREAM, route: "/premium", client: client(500, "boom"))
    status, _h, body = call(mw)
    assert_equal 502, status
    assert_includes body.join, "x402_platform_unavailable"
  end

  def test_no_proof_transport_error_fails_closed_502
    mw = X402::Middleware.new(DOWNSTREAM, route: "/premium", client: throwing_client)
    status, _h, body = call(mw)
    assert_equal 502, status
    assert_includes body.join, "x402_platform_unavailable"
  end

  # --- with proof → /verify ---

  def test_with_proof_verify_allowed_passthrough
    mw = X402::Middleware.new(DOWNSTREAM, route: "/premium",
                              client: client(200, '{"allowed":true}'))
    status, _h, body = call(mw, "HTTP_X_PAYMENT_NONCE" => "n-1",
                                "HTTP_X_PAYMENT" => '{"tx":"abc"}')
    assert downstream_ran?(status, body), "downstream handler should run"
  end

  def test_with_proof_verify_not_allowed_fails_closed_502
    mw = X402::Middleware.new(DOWNSTREAM, route: "/premium",
                              client: client(200, '{"allowed":false}'))
    status, = call(mw, "HTTP_X_PAYMENT_NONCE" => "n-1")
    assert_equal 502, status # verify 200 + not allowed ⇒ fail-closed
  end

  def test_with_proof_verify_402_relayed_as_402
    mw = X402::Middleware.new(DOWNSTREAM, route: "/premium",
                              client: client(402, '{"error":"payment_required"}'))
    status, _h, body = call(mw, "HTTP_X_PAYMENT_NONCE" => "n-1")
    assert_equal 402, status
    assert_includes body.join, "payment_required"
  end

  def test_with_proof_transport_error_fails_closed_502
    mw = X402::Middleware.new(DOWNSTREAM, route: "/premium", client: throwing_client)
    status, = call(mw, "HTTP_X_PAYMENT_NONCE" => "n-1")
    assert_equal 502, status
  end

  # --- compact JSON: signed bytes == sent bytes ---

  def test_compact_json_body_signed_bytes_equal_sent_bytes
    seen = {}
    transport = lambda do |_url, body, headers|
      seen[:body] = body
      seen[:sig] = headers["X-X402-Signature"]
      [200, '{"allowed":true}']
    end
    pc = X402::PlatformClient.new(CFG, transport: transport)
    pc.challenge("/premium", "GET")

    assert_equal '{"route":"/premium","method":"GET"}', seen[:body]
    assert_match(/\A[0-9a-f]{64}\z/, seen[:sig])
  end
end

File: spring/src/test/java/dev/x402/spring/MiddlewareTest.java
package dev.x402.spring;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.Map;

import org.junit.jupiter.api.Test;
import org.springframework.mock.web.MockFilterChain;
import org.springframework.mock.web.MockHttpServletRequest;
import org.springframework.mock.web.MockHttpServletResponse;

/**
 * Fail-closed flow tests using an injected fake {@link PlatformClient.Transport}
 * (no real HTTP) and Spring's servlet mocks.
 */
class MiddlewareTest {

    private static final X402Config CFG =
            new X402Config("key", "sk", "https://platform.test", "production");

    /** Fake transport returning a canned (status, body) for every call. */
    private static PlatformClient client(int status, String body) {
        PlatformClient.Transport t =
                (url, b, headers) -> new PlatformClient.Result(status, body);
        return new PlatformClient(CFG, t);
    }

    /** Transport that throws — simulates a network exception. */
    private static PlatformClient throwingClient() {
        PlatformClient.Transport t = (url, b, headers) -> {
            throw new RuntimeException("connection refused");
        };
        return new PlatformClient(CFG, t);
    }

    private MockHttpServletResponse run(X402Filter f, MockHttpServletRequest req,
            MockFilterChain chain) throws Exception {
        MockHttpServletResponse resp = new MockHttpServletResponse();
        f.doFilter(req, resp, chain);
        return resp;
    }

    @Test
    void noProof_challenge200_relayedAs402() throws Exception {
        X402Filter f = new X402Filter("/premium", "0.10",
                client(200, "{\"challenge\":\"pay-me\"}"));
        MockHttpServletRequest req = new MockHttpServletRequest("GET", "/premium");
        MockFilterChain chain = new MockFilterChain();

        MockHttpServletResponse resp = run(f, req, chain);

        assertEquals(402, resp.getStatus());
        assertTrue(resp.getContentAsString().contains("pay-me"));
        assertNull0(chain.getRequest()); // downstream NOT invoked
    }

    @Test
    void noProof_challenge404_relayedAs404() throws Exception {
        X402Filter f = new X402Filter("/premium", "0.10", client(404, "{}"));
        MockHttpServletRequest req = new MockHttpServletRequest("GET", "/premium");
        MockFilterChain chain = new MockFilterChain();

        MockHttpServletResponse resp = run(f, req, chain);

        assertEquals(404, resp.getStatus());
        assertTrue(resp.getContentAsString().contains("no_such_route"));
        assertNull0(chain.getRequest());
    }

    @Test
    void noProof_challenge500_failsClosed502() throws Exception {
        X402Filter f = new X402Filter("/premium", "0.10", client(500, "boom"));
        MockHttpServletRequest req = new MockHttpServletRequest("GET", "/premium");
        MockFilterChain chain = new MockFilterChain();

        MockHttpServletResponse resp = run(f, req, chain);

        assertEquals(502, resp.getStatus());
        assertTrue(resp.getContentAsString().contains("x402_platform_unavailable"));
        assertNull0(chain.getRequest());
    }

    @Test
    void noProof_transportException_failsClosed502() throws Exception {
        X402Filter f = new X402Filter("/premium", "0.10", throwingClient());
        MockHttpServletRequest req = new MockHttpServletRequest("GET", "/premium");
        MockFilterChain chain = new MockFilterChain();

        MockHttpServletResponse resp = run(f, req, chain);

        assertEquals(502, resp.getStatus());
        assertTrue(resp.getContentAsString().contains("x402_platform_unavailable"));
        assertNull0(chain.getRequest());
    }

    @Test
    void withProof_verifyAllowed_passthrough() throws Exception {
        X402Filter f = new X402Filter("/premium", "0.10",
                client(200, "{\"allowed\":true}"));
        MockHttpServletRequest req = new MockHttpServletRequest("GET", "/premium");
        req.addHeader("X-Payment-Nonce", "n-1");
        req.addHeader("X-Payment", "{\"tx\":\"abc\"}");
        MockFilterChain chain = new MockFilterChain();

        MockHttpServletResponse resp = run(f, req, chain);

        // Passthrough: chain ran, status untouched (200 default).
        assertEquals(200, resp.getStatus());
        assertTrue(chain.getRequest() != null, "downstream handler should run");
    }

    @Test
    void withProof_verifyDenied_402() throws Exception {
        X402Filter f = new X402Filter("/premium", "0.10",
                client(200, "{\"allowed\":false}"));
        MockHttpServletRequest req = new MockHttpServletRequest("GET", "/premium");
        req.addHeader("X-Payment-Nonce", "n-1");
        MockFilterChain chain = new MockFilterChain();

        MockHttpServletResponse resp = run(f, req, chain);

        assertEquals(502, resp.getStatus()); // verify status 200 + not allowed ⇒ fail-closed
        assertNull0(chain.getRequest());
    }

    @Test
    void withProof_verify402_relayedAs402() throws Exception {
        X402Filter f = new X402Filter("/premium", "0.10",
                client(402, "{\"error\":\"payment_required\"}"));
        MockHttpServletRequest req = new MockHttpServletRequest("GET", "/premium");
        req.addHeader("X-Payment-Nonce", "n-1");
        MockFilterChain chain = new MockFilterChain();

        MockHttpServletResponse resp = run(f, req, chain);

        assertEquals(402, resp.getStatus());
        assertTrue(resp.getContentAsString().contains("payment_required"));
        assertNull0(chain.getRequest());
    }

    @Test
    void withProof_verifyTransportError_failsClosed502() throws Exception {
        X402Filter f = new X402Filter("/premium", "0.10", throwingClient());
        MockHttpServletRequest req = new MockHttpServletRequest("GET", "/premium");
        req.addHeader("X-Payment-Nonce", "n-1");
        MockFilterChain chain = new MockFilterChain();

        MockHttpServletResponse resp = run(f, req, chain);

        assertEquals(502, resp.getStatus());
        assertNull0(chain.getRequest());
    }

    @Test
    void compactJsonBody_signedBytesEqualSentBytes() {
        // Capture the body the transport receives and re-derive the signature
        // from those exact bytes — proves no serializer drift.
        final String[] seen = new String[2];
        PlatformClient.Transport t = (url, body, headers) -> {
            seen[0] = body;
            seen[1] = headers.get("X-X402-Signature");
            return new PlatformClient.Result(200, "{\"allowed\":true}");
        };
        PlatformClient pc = new PlatformClient(CFG, t);
        pc.challenge("/premium", "GET");

        assertEquals("{\"route\":\"/premium\",\"method\":\"GET\"}", seen[0]);
        // Signature header must be a 64-char lowercase hex over those bytes.
        assertTrue(seen[1].matches("^[0-9a-f]{64}$"));
    }

    private static void assertNull0(Object o) {
        assertFalse(o != null, "downstream chain must NOT have been invoked");
    }
}
